### ISSUE TITLE
Compiling Troubles (not-code related)

### ISSUE THREAD

**Issue by crollinson** [EDmodel/ED2#153]
I'm trying to compile a stable branch of [ED](https://github.com/crollinson/ED2/tree/paleon_region_v1) on University of Arizona's Unix-based HPC.  Their directions strongly recommend using Intel-specific compilers, and I've tried following the directions they provide for [compiling with OpenMP](http://rc.arizona.edu/hpc-htc/using-systems/compliers-and-options).  However, with this [include.mk.opt file](https://github.com/crollinson/ED_PalEON/blob/UA-HPC/MIP2_Region/0_setup/include.mk.opt), I receive the following errors: 

```
ed_state_vars.F90(11181): error #5102: Cannot open include file 'mpif.h'
      include 'mpif.h'
--------------^
ed_state_vars.F90(11253): error #6404: This name does not have a type, and must have an explicit type.   [MPI_INTEGER]
               call MPI_Send(ping,1,MPI_INTEGER,sendnum,94,MPI_COMM_WORLD,err)
------------------------------------^
ed_state_vars.F90(11253): error #6404: This name does not have a type, and must have an explicit type.   [MPI_COMM_WORLD]
               call MPI_Send(ping,1,MPI_INTEGER,sendnum,94,MPI_COMM_WORLD,err)
-----------------------------------------------------------^
ed_state_vars.F90(11200): error #6279: A specification expression object must be a dummy argument, a COMMON block object, or an object accessible through host or use association   [MPI_STATUS_SIZE]
      integer          , dimension(MPI_STATUS_SIZE)            :: status
-----------------------------------^
ed_state_vars.F90(11200): error #6404: This name does not have a type, and must have an explicit type.   [MPI_STATUS_SIZE]
      integer          , dimension(MPI_STATUS_SIZE)            :: status
-----------------------------------^

```

Has anybody seen these and know the fix?  I tried using different compilers including loading openmpi/1.6.4 and using mpif90 and mpicc, but that doesn't even begin to compile.

I've played around with the settings quite a bit, but haven't had any luck.  I suspect this is a location-specific problem since I've successfully compiled ED in several other locations, but if anybody has any tips, they'd be greatly appreciated!

**Comment by mpaiao** [comment]
Which error does it show when you tried to compile with mpif90/mpicc?  If you use the mpi compilers, leave MPI_PATH, PAR_INCS and PAR_LIBS unset.  Otherwise, if you use ifort/icc, then you must either set MPI_PATH,PAR_INCS and PAR_LIBS, or disable mpi altogether (i.e. leave PAR_DEFS blank).  I'm not 100% sure but I think the OMP/-openmp may still work even if PAR_DEFS is unset, it's just not going to run multiple polygons in parallel.

By the way, I noticed that the uncommented F_OPTS and MPI_PATH variables are written with lowercase letters, but I think Makefile is case sensitive.

**Comment by crollinson** [comment]
I got through the previous error by using mpiicc and mpiifort, but am now running into hdf5 trouble:

cp -f ../../src/memory/hdf5_coms.F90 hdf5_coms.F90
mpiifort -c -DUSE_INTERF=1 -DUSENC=0 -DPC_LINUX1 -DUSE_HDF5=1 -DUSE_COLLECTIVE_MPIO=1 -DUSE_MPIWTIME=1   -I../../src/include -I/uaopt/hdf5/1.8.12/include   -DRAMS_MPI -I/uaopt/hdf5/1.8.12/include hdf5_coms.F90
hdf5_coms.F90(3): error #7002: Error in opening the compiled module file.  Check INCLUDE paths.   [HDF5]
  use hdf5
-------^
hdf5_coms.F90(5): error #6683: A kind type parameter must be a compile-time constant.   [HID_T]
   integer(HID_T) :: dsetid_f,dspaceid_f,fileid_f,prpid_f
-----------^
hdf5_coms.F90(6): error #6683: A kind type parameter must be a compile-time constant.   [HID_T]
   integer(HID_T) :: plist_id,file_id,dset_id,filespace,memspace,dspace_id
-----------^
hdf5_coms.F90(8): error #6683: A kind type parameter must be a compile-time constant.   [HSIZE_T]
   integer(HSIZE_T),dimension(6) :: globdims,chnkdims,chnkoffs
-----------^
hdf5_coms.F90(9): error #6683: A kind type parameter must be a compile-time constant.   [HSIZE_T]
   integer(HSIZE_T),dimension(6) :: cnt,stride
-----------^
hdf5_coms.F90(11): error #6683: A kind type parameter must be a compile-time constant.   [HSIZE_T]
   integer(HSIZE_T),dimension(6) :: memsize,memdims,memoffs !EQUIV IN MEM
-----------^
hdf5_coms.F90(13): error #6683: A kind type parameter must be a compile-time constant.   [HID_T]
   integer(HID_T)  :: datatype_id ! Datatype identifier
-----------^
hdf5_coms.F90(14): error #6683: A kind type parameter must be a compile-time constant.   [SIZE_T]
   integer(SIZE_T) :: setsize  ! Datatype size
-----------^
compilation aborted for hdf5_coms.F90 (code 1)



The hdf5 part of my include.mk.opt file looks like this: 
HDF5HOME=/uaopt/hdf5/1.8.12
USE_HDF5=1
HDF5_INCS=-I${HDF5HOME}/include
HDF5_LIBS=-L${HDF5HOME}/lib -lhdf5 -lm -lhdf5_fortran -lhdf5 -lhdf5_hl -lz
USE_COLLECTIVE_MPIO=1



Pretty sure this is a UA-specific problem, but if folks have any suggestions, that'd be great.

**Comment by mdietze** [comment]
Key error seems to be "Error in opening the compiled module file. Check INCLUDE paths."

Did you verify that hdf5 exists at HDF5HOME and you don't need to do any sort of module load? Have others on that system successfully compiled against that library? Are there simple hdf5 tests that can be run to verify the package was installed correctly (including fortran support)?

**Comment by mpaiao** [comment]
Also, I think the order of the libraries matter here. Typically libz is not in the same path as the hdf5 libraries.  Does it go through if you try:
```Makefile
HDF5_LIBS=-lz -L${HDF5HOME}/lib -lhdf5 -lhdf5_fortran -lhdf5_hl
```
or
```Makefile
HDF5_LIBS=-L$(PATH_WHERE_LIBZ_IS_LOCATED) -lz -L${HDF5HOME}/lib -lhdf5 -lhdf5_fortran  -lhdf5_hl
```
(I don't compile with libm, and I think it's not needed for ED)

**Comment by crollinson** [comment]
after digging a bit more and comparing the Arizona library with BU, it looks like they do not have hdf5 installed with fortran support.  I'm assuming there's no work around until I can get IT to do something about it.

(I do have the correct HDF5HOME path and already tried switching up the order of the libraries.)

**Comment by crollinson** [comment]
Making progress, but no solution yet.  Arizona added the hdf5 libraries to hdf5/1.10.0, but they don't seem to be working.  Similar error, with a slightly different description:

mpiifort -c -DUSE_INTERF=1 -DUSENC=0 -DPC_LINUX1 -DUSE_HDF5=1 -DUSE_COLLECTIVE_MPIO=1 -DUSE_MPIWTIME=1  -I/uaopt/intel/2012.0.032/impi/4.0.3.008/intel64//include -I../../src/include -I/uaopt/hdf5/1.10.0/include   -DRAMS_MPI -I/uaopt/hdf5/1.10.0/include hdf5_coms.F90

hdf5_coms.F90(3): error #7013: This module file was not generated by any release of this compiler.   [HDF5]

   use hdf5
-------^

[and then all the same errors as above]

Any suggestions on whether this is an issue with the version of the hdf5 library versus a missing connection between the hdf5 and mpiifort (intel-mpi)?  I updated the HDF5HOME path and tried all the different ways of specifying the library paths that Marcos suggested and that didn't change anything.

**Comment by DanielNScott** [comment]
```This module file was not generated by any release of this compiler.``` would seem to imply that mpiifort is using a different (potentially earlier) version of intel fortran than the compiler HDF5 was installed with[

**Comment by crollinson** [comment]
Something about hdf5 appears to still be giving me trouble across multiple systems now.  If I use mpif90 and mpicc I am able to compile the hdf5_coms module, but then things bonk here in hdf5_utils: 

cp -f ../../src/utils/hdf5_utils.F90 hdf5_utils.F90
mpif90 -c -DUSE_INTERF=1 -DUSENC=0 -DPC_LINUX1 -DUSE_HDF5=1 -DUSE_COLLECTIVE_MPIO=1 -DUSE_MPIWTIME=1   -I../../src/include -I/uaopt/hdf5/1.10.0/include   -DRAMS_MPI -I/uaopt/hdf5/1.10.0/include hdf5_utils.F90
hdf5_utils.F90:469.30:

  call h5dget_type_f(dsetid_f,type_id,hdferr)
                              1
Error: Type mismatch in argument 'datatype_id' at (1); passed INTEGER(4) to INTEGER(8)
make[1]: *** [hdf5_utils.o] Error 1
make[1]: Leaving directory `/gsfs2/home/u7/crollinson/ED2/ED/build/bin'
make: *** [all] Error 2


On a completely unrelated system I'm able to get the program to compile, but then we keep getting a segfault error when it tries to load the land/sea mask.  It's the same kind of error and at the same point that we get at BU if it's compiled for parallel, but we're not logged onto the parallel computing system.  My collaborator assures me we're logged into the appropriate system and suspect it too is a hdf5-related problem.

**Comment by mpaiao** [comment]
Regarding the land/sea mask segmentation violation, have you tried to add
```bash
ulimit -s unlimited
```
to your .bashrc?

@lucianainpa was having a similar problem in her machine, and that solved the problem.

**Comment by crollinson** [comment]
Finally got my original problem fixed!  Changing the definition of type_id from an integer to HID_T in line 435 of hdf5_utils.F90 worked (see [commit](https://github.com/crollinson/ED2/commit/48b9e8132b992460aa8afeca222fd4c12f01cfe4) )

The BU support guy suggested pushing this change to the mainline since it might help solve problems from @ashiklom in #159

**Comment by fabeit** [comment]
@crollinson I am having the same problem you described in your first post, I am compiling with kind A,B,C, all of them get stuck at the same point. I am able to compile no problem in E. I have tried using mpif90 and mpicc with no luck, I have tried with both mpi paths blank or specified, same results.

**Comment by fabeit** [comment]
nevermind, issue resolved.

**Comment by aruhela** [comment]
> 
> 
> after digging a bit more and comparing the Arizona library with BU, it looks like they do not have hdf5 installed with fortran support. I'm assuming there's no work around until I can get IT to do something about it.
> 
> (I do have the correct HDF5HOME path and already tried switching up the order of the libraries.)

Thanks for the solution. I am fortunate to go throgh this issue.
